classes/conversion.php

```diff
@@ -613,37 +665,93 @@ private function get_transcode_files(\stdClass $conversionrecord, $handler=null)
+        foreach ($availableobjects->get('Contents') as $availableobject) {
+            $filename = basename($availableobject['Key']);
             $filerecord = array(
                 'contextid' => 1, // Put files in the site level context as they aren't associated with a specific context.
                 'component' => 'local_smartmedia',
                 'filearea' => 'media',
                 'itemid' => 0,
                 'filepath' => '/' . $conversionrecord->contenthash . '/conversions/',
-                'filename' => basename($availableobject['Key'])
-
+                'filename' => $filename,
             );
 
             $downloadparams = array(
-                    'Bucket' => $this->config->s3_output_bucket, // Required.
-                    'Key' => $availableobject['Key'], // Required.
+                'Bucket' => $this->config->s3_output_bucket, // Required.
+                'Key' => $availableobject['Key'], // Required.
             );
 
             $getobject = $s3client->getObject($downloadparams);
+            $filecontent = $getobject->get('Body');
 
             $tmpfile = tmpfile();
-            fwrite($tmpfile, $getobject['Body']);
+            fwrite($tmpfile, $filecontent);
             $tmppath = stream_get_meta_data($tmpfile)['uri'];
 
-            $fs->create_file_from_pathname($filerecord, $tmppath);
+            // The playlist files (including iframe playlists) created in s3 transcoding contain
+            // relative file paths to Variant Streams for adaptive bitsteaming media, these need to be amended
+            // to have Moodle plugin filepaths.
+            if ($this->is_file_playlist($filename)) {
+                $newfile = $this->replace_playlist_urls_with_pluginfile_urls($tmpfile, $conversionrecord->contenthash);
+                $tmppath = stream_get_meta_data($newfile)['uri'];
+            }
+
+            $trancodedfile = $fs->create_file_from_pathname($filerecord, $tmppath);
             fclose($tmpfile);
+            if (isset($newfile)) {
+                fclose($newfile);
+                unset($newfile); // Make sure we aren't overriding the same resource every time.
+            }
+            $transcodedfiles[] = $trancodedfile;
         }
-
+        return $transcodedfiles;
         // TODO: Also remove files from AWS.
     }
 
+    /**
+     * Replace relative urls in a media playlist with pluginfile urls so the playlist may serve files in Moodle.
+     *
+     * @param resource $filehandle the handle for the file to replace urls in.
+     * @param string $contenthash the content hash for conversion to search for and replace.
+     *
+     * @return resource $newfile file handle for a new file created with amended playlist data/
+     */
+    private function replace_playlist_urls_with_pluginfile_urls($filehandle, string $contenthash) {
+        global $CFG;
+
+        rewind($filehandle);
+        $newfile = tmpfile();
```

Comment:
I'm pretty sure Moodle has a tempfile function. We should probably be using that